File: jormungandr/test/optimization/linear_problem_test.py
import pytest

from jormungandr.autodiff import ExpressionType
from jormungandr.optimization import ExitStatus, Problem


def test_maximize():
    problem = Problem()

    x = problem.decision_variable()
    x.set_value(1.0)

    y = problem.decision_variable()
    y.set_value(1.0)

    problem.maximize(50 * x + 40 * y)

    problem.subject_to(x + 1.5 * y <= 750)
    problem.subject_to(2 * x + 3 * y <= 1500)
    problem.subject_to(2 * x + y <= 1000)
    problem.subject_to(x >= 0)
    problem.subject_to(y >= 0)

    assert problem.cost_function_type() == ExpressionType.LINEAR
    assert problem.equality_constraint_type() == ExpressionType.NONE
    assert problem.inequality_constraint_type() == ExpressionType.LINEAR

    assert problem.solve(diagnostics=True) == ExitStatus.SUCCESS

    assert x.value() == pytest.approx(375.0, abs=1e-6)
    assert y.value() == pytest.approx(250.0, abs=1e-6)


def test_free_variable():
    problem = Problem()

    x = problem.decision_variable(2)
    x[0].set_value(1.0)
    x[1].set_value(2.0)

    problem.subject_to(x[0] == 0)

    assert problem.cost_function_type() == ExpressionType.NONE
    assert problem.equality_constraint_type() == ExpressionType.LINEAR
    assert problem.inequality_constraint_type() == ExpressionType.NONE

    assert problem.solve(diagnostics=True) == ExitStatus.SUCCESS

    assert x[0].value() == pytest.approx(0.0, abs=1e-6)
    assert x[1].value() == pytest.approx(2.0, abs=1e-6)

File: test/src/optimization/linear_problem_test.cpp
// Copyright (c) Sleipnir contributors

#include <catch2/catch_template_test_macros.hpp>
#include <catch2/catch_test_macros.hpp>
#include <sleipnir/optimization/problem.hpp>

#include "catch_matchers.hpp"
#include "catch_string_converters.hpp"
#include "scalar_types_under_test.hpp"

TEMPLATE_TEST_CASE("Problem - Maximize", "[Problem]", SCALAR_TYPES_UNDER_TEST) {
  using T = TestType;

  slp::Problem<T> problem;

  auto x = problem.decision_variable();
  x.set_value(T(1));

  auto y = problem.decision_variable();
  y.set_value(T(1));

  problem.maximize(T(50) * x + T(40) * y);

  problem.subject_to(x + T(1.5) * y <= T(750));
  problem.subject_to(T(2) * x + T(3) * y <= T(1500));
  problem.subject_to(T(2) * x + y <= T(1000));
  problem.subject_to(x >= T(0));
  problem.subject_to(y >= T(0));

  CHECK(problem.cost_function_type() == slp::ExpressionType::LINEAR);
  CHECK(problem.equality_constraint_type() == slp::ExpressionType::NONE);
  CHECK(problem.inequality_constraint_type() == slp::ExpressionType::LINEAR);

  CHECK(problem.solve({.diagnostics = true}) == slp::ExitStatus::SUCCESS);

  CHECK_THAT(x.value(), WithinAbs(T(375), T(1e-6)));
  CHECK_THAT(y.value(), WithinAbs(T(250), T(1e-6)));
}

TEMPLATE_TEST_CASE("Problem - Free variable", "[Problem]",
                   SCALAR_TYPES_UNDER_TEST) {
  using T = TestType;

  slp::Problem<T> problem;

  auto x = problem.decision_variable(2);
  x[0].set_value(T(1));
  x[1].set_value(T(2));

  problem.subject_to(x[0] == T(0));

  CHECK(problem.cost_function_type() == slp::ExpressionType::NONE);
  CHECK(problem.equality_constraint_type() == slp::ExpressionType::LINEAR);
  CHECK(problem.inequality_constraint_type() == slp::ExpressionType::NONE);

  CHECK(problem.solve({.diagnostics = true}) == slp::ExitStatus::SUCCESS);

  CHECK_THAT(x[0].value(), WithinAbs(T(0), T(1e-6)));
  CHECK_THAT(x[1].value(), WithinAbs(T(2), T(1e-6)));
}
